app/domain/authentication/authn_jwt/authenticator.rb

```diff
@@ -0,0 +1,141 @@
+
+      def call
+        validate_and_decode_token
+        get_jwt_identity_from_request
+        validate_user_has_access_to_webservice
+        validate_origin
+        validate_restrictions
+        audit_success
+        @logger.debug(LogMessages::Authentication::AuthnJwt::JwtAuthenticationPassed.new)
+        new_token
+      rescue => e
+        audit_failure(e)
+        raise e
+      end
+
+      private
+
+      def validate_and_decode_token
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingValidateAndDecodeToken.new)
+        @jwt_configuration.validate_and_decode_token
+        @logger.debug(LogMessages::Authentication::AuthnJwt::ValidateAndDecodeTokenPassed.new)
+      end
+
+      def get_jwt_identity_from_request
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingGetJwtIdentity.new)
+        jwt_identity
+        @logger.info(LogMessages::Authentication::AuthnJwt::FoundJwtIdentity.new(jwt_identity))
+      end
+
+      def jwt_identity
+        @jwt_identity ||= @jwt_configuration.jwt_identity
+      end
+
+      def validate_user_has_access_to_webservice
+        @validate_role_can_access_webservice.(
+          webservice: webservice,
+          account: account,
+          user_id: jwt_identity,
+          privilege: PRIVILEGE_AUTHENTICATE
+        )
+      end
+
+      def validate_origin
+        @validate_origin.(
+          account: account,
+          username: jwt_identity,
+          client_ip: client_ip
+        )
+      end
+
+      def validate_restrictions
+        @logger.debug(LogMessages::Authentication::AuthnJwt::CallingValidateRestrictions.new)
+        @jwt_configuration.validate_restrictions
+        @logger.debug(LogMessages::Authentication::AuthnJwt::ValidateRestrictionsPassed.new)
+      end
+
+      def audit_success
+        @audit_log.log(
+          ::Audit::Event::Authn::Authenticate.new(
+            authenticator_name: authenticator_name,
+            service: webservice,
+            role_id: audit_role_id,
+            client_ip: client_ip,
+            success: true,
+            error_message: nil
+          )
+        )
+      end
+
+      def audit_failure(err)
+        @audit_log.log(
+          ::Audit::Event::Authn::Authenticate.new(
+            authenticator_name: authenticator_name,
+            service: webservice,
+            role_id: audit_role_id,
+            client_ip: client_ip,
+            success: false,
+            error_message: err.message
+          )
+        )
+      end
+
+      def identity_role
+        @identity_role ||= @role_class.by_login(
+          jwt_identity,
+          account: account
+        )
+      end
+
+      # If there is no jwt identity so role and username are nil
+      def audit_role_id
+        return @audit_role_id if @audit_role_id
+        if @jwt_identity
+          role = identity_role
+          username = jwt_identity
+        end
+        @audit_role_id ||= @role_id_class.new(
```

Comment:
no need for memoization as you already have a guard above
```suggestion
        @audit_role_id = @role_id_class.new(
```

Reply:
Done